File: simulations-java/simulations/circuit-construction-kit/src/edu/colorado/phet/circuitconstructionkit/model/mna/TimestepSubdivisions.java
package edu.colorado.phet.circuitconstructionkit.model.mna;

import java.util.ArrayList;

/**
 * TimestepSubdivisions updates a state over an interval dt by (potentially) subdividing it into smaller regions, potentially with different lengths.
 * To select the (sub) time step for each iteration, the difference between an update of h and two updates of h/2 are performed.
 * If the error between the h vs. 2x(h/2) states is within the tolerated threshold, the time step is accepted.
 * See #2241
 *
 * @author Sam Reid
 * @param <A> the state type that is being updated
 */
public class TimestepSubdivisions<A> {
    private double errorThreshold;//threshold for determining whether 2 states are similar enough; any error less than errorThreshold will be tolerated.
    private double minDT;//lowest possible value for DT, independent of how the error scales with reduced time step

    public TimestepSubdivisions(double errorThreshold, double minDT) {
        this.errorThreshold = errorThreshold;
        this.minDT = minDT;
    }

    public static interface Steppable<A> {
        double distance(A a, A b);

        A update(A a, double dt);
    }

    public static class State<A> {
        double dt;
        A state;

        public State(double dt, A state) {
            this.state = state;
            this.dt = dt;
        }
    }

    public static class Result<A> {
        ArrayList<State<A>> states;

        public Result(ArrayList<State<A>> states) {
            this.states = states;
        }

        public A getFinalState() {
            return states.get(states.size() - 1).state;
        }
    }

    public Result<A> stepInTimeWithHistory(A originalState, Steppable<A> steppable, double dt) {
        A state = originalState;
        double elapsed = 0.0;
        ArrayList<State<A>> states = new ArrayList<State<A>>();
        while (elapsed < dt) {
            double seedValue = states.size() > 0 ? states.get(states.size() - 1).dt : dt;//use the last obtained dt as a starting value, if possible
            double subdivisionDT = getTimestep(state, steppable, seedValue);
            System.out.println("selected subdivisionDT = " + subdivisionDT);
            if (subdivisionDT + elapsed > dt) subdivisionDT = dt - elapsed; // don't exceed max allowed dt
            state = steppable.update(state, subdivisionDT);
            states.add(new State<A>(subdivisionDT, state));
            elapsed = elapsed + subdivisionDT;
        }
        return new Result<A>(states);
    }

    public A stepInTime(A originalState, Steppable<A> steppable, double dt) {
        return stepInTimeWithHistory(originalState, steppable, dt).getFinalState();
    }

    protected double getTimestep(A state, Steppable<A> steppable, double dt) {
        //store the previously used DT and try it first, then to increase it when possible.
        if (dt < minDT) {
            System.out.println("Time step too small");
            return minDT;
        } else if (errorAcceptable(state, steppable, dt * 2))
            return dt * 2; //only increase by one factor if this exceeds the totalDT, it will be cropped later
        else if (errorAcceptable(state, steppable, dt)) return dt * 2;
        else return getTimestep(state, steppable, dt / 2);
    }

    protected boolean errorAcceptable(A state, Steppable<A> steppable, double dt) {
        A a = steppable.update(state, dt);
        A b1 = steppable.update(state, dt / 2);
        A b2 = steppable.update(b1, dt / 2);
        return steppable.distance(a, b2) < errorThreshold;
    }
}

File: branches/test-branch-44224/simulations-java/simulations/circuit-construction-kit/scala-src/edu/colorado/phet/cckscala/tests/TimestepSubdivisions.scala
package edu.colorado.phet.cckscala.tests

import collection.mutable.ArrayBuffer

trait Steppable[A] {
  def distance(a: A, b: A): Double

  def update(a: A, dt: Double): A
}

case class Result[A](states:Seq[(Double,A)],state:A)

class TimestepSubdivisions(errorThreshold: Double) {
  def stepInTimeWithHistory[A](originalState: A, steppable: Steppable[A], dt: Double) = {
    var state = originalState
    var elapsed = 0.0
    val states = new ArrayBuffer[(Double,A)]
    while (elapsed < dt) {
      //println("elapsed = "+elapsed +", dt = "+dt)
      val seedValue = if (states.length>0) states(states.length-1)._1 else dt
      var subdivisionDT = getTimestep(state, steppable, seedValue)
      if (subdivisionDT + elapsed > dt) subdivisionDT = dt - elapsed // don't exceed max allowed dt
      state = steppable.update(state, subdivisionDT)
      states += ((subdivisionDT,state))
      elapsed = elapsed + subdivisionDT
    }
    Result(states.toList,state)
  }

  def stepInTime[A](originalState: A, steppable: Steppable[A], dt: Double) = stepInTimeWithHistory(originalState,steppable,dt).state

  def getTimestep[A](state: A, steppable: Steppable[A], dt: Double): Double = {
    //store the previously used DT and try it first, then to increase it when possible.
    if (dt < 1E-8) {
      println("Time step too small")
      dt
    }
    else if (errorAcceptable(state, steppable, dt * 2))
      dt * 2 //only increase by one factor if this exceeds the totalDT, it will be cropped later
    else if (errorAcceptable(state, steppable, dt)) dt * 2
    else getTimestep(state, steppable, dt / 2)
  }

  def errorAcceptable[A](state: A, steppable: Steppable[A], dt: Double): Boolean = {
    if (dt < 1E-6)
      true
    else {
      val a = steppable.update(state, dt)
      val b1 = steppable.update(state, dt / 2)
      val b2 = steppable.update(b1, dt / 2)
      steppable.distance(a, b2) < errorThreshold
    }
  }
}